src/core/seqner.rs

```diff
@@ -0,0 +1,268 @@
+use lazy_static::lazy_static;
+
+use crate::core::matter::{tables as matter, Matter};
+use crate::error::{err, Error, Result};
+
+#[derive(Debug, Clone, PartialEq)]
+pub struct Seqner {
+    raw: Vec<u8>,
+    code: String,
+    size: u32,
+}
+
+impl Default for Seqner {
+    fn default() -> Self {
+        Seqner { raw: vec![], code: matter::Codex::Salt_128.to_string(), size: 0 }
+    }
+}
+
+fn validate_code(code: &str) -> Result<()> {
+    lazy_static! {
+        static ref CODES: Vec<&'static str> = vec![matter::Codex::Salt_128,];
+    }
+
+    if !CODES.contains(&code) {
+        return err!(Error::UnexpectedCode(code.to_string()));
+    }
+
+    Ok(())
+}
+
+impl Seqner {
+    #[allow(clippy::too_many_arguments)]
+    pub fn new(
+        sn: Option<u128>,
+        snh: Option<&str>,
+        code: Option<&str>,
+        raw: Option<&[u8]>,
+        qb64b: Option<&mut Vec<u8>>,
+        qb64: Option<&str>,
+        qb2: Option<&mut Vec<u8>>,
+        strip: Option<bool>,
+    ) -> Result<Self> {
+        let buf;
+        let raw: Option<&[u8]> =
+            if raw.is_none() && qb64b.is_none() && qb64.is_none() && qb2.is_none() {
+                let sn = if let Some(sn) = sn {
+                    sn
+                } else if let Some(snh) = snh {
+                    u128::from_str_radix(snh, 16)?
+                } else {
+                    0_u128
+                };
+                buf = sn.to_be_bytes();
+                Some(&buf)
+            } else {
+                raw
+            };
+        let seqner: Seqner = Matter::new(code, raw, qb64b, qb64, qb2, strip)?;
+        validate_code(&seqner.code)?;
+        Ok(seqner)
+    }
+
+    pub fn sn(&self) -> Result<u128> {
+        Ok(u128::from_be_bytes(self.raw[0..16].try_into()?))
+    }
+
+    pub fn snh(&self) -> Result<String> {
+        Ok(format!("{:x}", self.sn()?))
```

Comment:
Will this strip leading 0s?

Reply:
Yeah, please add tests for that case. I don't see any

Reply:
Just see what `.snh()` outputs when sn == 0.

Reply:
I suspect you may need something like
`Ok(format!("{:032x}", self.sn()?))`

Reply:
Sorry I found the test. So it is expected that it will strip 0s.

Reply:
Yes, Python implementation is always returns `snh` as a stripped lower-case hex string.